{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "24feea3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import GEOparse\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import torch\n",
    "from anndata import AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7a32e4b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_root = '/work/hdd/bbjr/mallina1/data/mb-ml-dev-vm/data/GSE153807/tsvs'\n",
    "\n",
    "fnames = ['GSM4654467_Nuc-RM101-2.raw.tsv', \n",
    "          'GSM4654469_Nuc-RM102-1.raw.tsv', 'GSM4654468_Nuc-RM102-2.raw.tsv', \n",
    "          'GSM4654470_Nuc-RM77-1.raw.tsv', 'GSM4654471_Nuc-RM77-2.raw.tsv',\n",
    "          'GSM4654472_Nuc-RM95-1.raw.tsv', 'GSM4654473_Nuc-RM95-2.raw.tsv']\n",
    "\n",
    "sex_per_fname = ['Female', 'Female', 'Female', 'Male', 'Male', 'Female', 'Female']\n",
    "\n",
    "gene_info_path = '/work/hdd/bbjr/mallina1/data/mb-ml-dev-vm/gene_info/gene_info.tsv'\n",
    "ontology_infos_path = '/work/hdd/bbjr/mallina1/data/mb-ml-dev-vm/ontology_infos.pt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "344605b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1967536/2574708238.py:1: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  ontology_infos = torch.load(ontology_infos_path)\n"
     ]
    }
   ],
   "source": [
    "ontology_infos = torch.load(ontology_infos_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6538fa8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'names': ['PATO:0000384', 'PATO:0000383'], 'labels': ['male', 'female']}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ontology_infos['sex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e3e7b7f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['embryonic stage',\n",
       " 'prenatal stage',\n",
       " 'organogenesis stage',\n",
       " 'Carnegie stage 12',\n",
       " 'Carnegie stage 13',\n",
       " 'Carnegie stage 14',\n",
       " 'Carnegie stage 16',\n",
       " 'Carnegie stage 17',\n",
       " 'Carnegie stage 18',\n",
       " 'Carnegie stage 19',\n",
       " 'Carnegie stage 20',\n",
       " 'Carnegie stage 21',\n",
       " 'Carnegie stage 22',\n",
       " 'Carnegie stage 23',\n",
       " 'fetal stage',\n",
       " '9th week post-fertilization stage',\n",
       " 'third LMP month stage',\n",
       " '10th week post-fertilization stage',\n",
       " '11th week post-fertilization stage',\n",
       " '12th week post-fertilization stage',\n",
       " 'fourth LMP month stage',\n",
       " '13th week post-fertilization stage',\n",
       " '14th week post-fertilization stage',\n",
       " '15th week post-fertilization stage',\n",
       " '16th week post-fertilization stage',\n",
       " 'fifth LMP month stage',\n",
       " '17th week post-fertilization stage',\n",
       " '18th week post-fertilization stage',\n",
       " '19th week post-fertilization stage',\n",
       " '20th week post-fertilization stage',\n",
       " 'sixth LMP month stage',\n",
       " '21st week post-fertilization stage',\n",
       " '22nd week post-fertilization stage',\n",
       " '23rd week post-fertilization stage',\n",
       " '25th week post-fertilization stage',\n",
       " 'seventh LMP month stage',\n",
       " '26th week post-fertilization stage',\n",
       " 'eighth LMP month stage',\n",
       " '31st week post-fertilization stage',\n",
       " 'ninth LMP month stage',\n",
       " 'immature stage',\n",
       " 'child stage',\n",
       " 'newborn stage',\n",
       " 'infant stage',\n",
       " '2-5 year-old child stage',\n",
       " '6-12 year-old child stage',\n",
       " 'adolescent stage',\n",
       " 'adult stage',\n",
       " 'early adulthood stage',\n",
       " 'young adult stage',\n",
       " '25-44 year-old stage',\n",
       " 'late adulthood stage',\n",
       " 'middle aged stage',\n",
       " 'aged stage',\n",
       " '65-79 year-old stage',\n",
       " '80 year-old and over stage',\n",
       " 'late adult stage',\n",
       " '79-year-old stage',\n",
       " '60-79 year-old stage',\n",
       " '2-year-old stage',\n",
       " '2-4 year-old child stage',\n",
       " '3-year-old stage',\n",
       " '4-year-old stage',\n",
       " '5-year-old stage',\n",
       " 'juvenile stage (5-14 yo)',\n",
       " '6-year-old stage',\n",
       " '7-year-old stage',\n",
       " '8-year-old stage',\n",
       " '9-year-old stage',\n",
       " '10-year-old stage',\n",
       " '11-year-old stage',\n",
       " '12-year-old stage',\n",
       " '13-year-old stage',\n",
       " '14-year-old stage',\n",
       " '15-year-old stage',\n",
       " '15-19 year-old',\n",
       " '16-year-old stage',\n",
       " '17-year-old stage',\n",
       " '18-year-old stage',\n",
       " '19-year-old stage',\n",
       " '20-year-old stage',\n",
       " 'third decade stage',\n",
       " '21-year-old stage',\n",
       " '22-year-old stage',\n",
       " '23-year-old stage',\n",
       " '24-year-old stage',\n",
       " '25-year-old stage',\n",
       " '26-year-old stage',\n",
       " '27-year-old stage',\n",
       " '28-year-old stage',\n",
       " '29-year-old stage',\n",
       " '30-year-old stage',\n",
       " 'fourth decade stage',\n",
       " '31-year-old stage',\n",
       " '32-year-old stage',\n",
       " '33-year-old stage',\n",
       " '34-year-old stage',\n",
       " '35-year-old stage',\n",
       " '36-year-old stage',\n",
       " '37-year-old stage',\n",
       " '38-year-old stage',\n",
       " '39-year-old stage',\n",
       " '40-year-old stage',\n",
       " 'fifth decade stage',\n",
       " '41-year-old stage',\n",
       " '42-year-old stage',\n",
       " '43-year-old stage',\n",
       " '44-year-old stage',\n",
       " '45-year-old stage',\n",
       " '46-year-old stage',\n",
       " '47-year-old stage',\n",
       " '48-year-old stage',\n",
       " '49-year-old stage',\n",
       " '50-year-old stage',\n",
       " 'sixth decade stage',\n",
       " '51-year-old stage',\n",
       " '52-year-old stage',\n",
       " '53-year-old stage',\n",
       " '54-year-old stage',\n",
       " '55-year-old stage',\n",
       " '56-year-old stage',\n",
       " '57-year-old stage',\n",
       " '58-year-old stage',\n",
       " '59-year-old stage',\n",
       " '60-year-old stage',\n",
       " 'seventh decade stage',\n",
       " '61-year-old stage',\n",
       " '62-year-old stage',\n",
       " '63-year-old stage',\n",
       " '64-year-old stage',\n",
       " '65-year-old stage',\n",
       " '66-year-old stage',\n",
       " '67-year-old stage',\n",
       " '68-year-old stage',\n",
       " '69-year-old stage',\n",
       " '70-year-old stage',\n",
       " 'eighth decade stage',\n",
       " '71-year-old stage',\n",
       " '72-year-old stage',\n",
       " '73-year-old stage',\n",
       " '74-year-old stage',\n",
       " '75-year-old stage',\n",
       " '76-year-old stage',\n",
       " '77-year-old stage',\n",
       " '78-year-old stage',\n",
       " '1-month-old stage',\n",
       " '2-month-old stage',\n",
       " 'infant stage',\n",
       " '3-month-old stage',\n",
       " '4-month-old stage',\n",
       " '5-month-old stage',\n",
       " '6-month-old stage',\n",
       " '7-month-old stage',\n",
       " '8-month-old stage',\n",
       " '9-month-old stage',\n",
       " '10-month-old stage',\n",
       " '11-month-old stage',\n",
       " '12-month-old stage',\n",
       " '1-year-old stage',\n",
       " '14-month-old stage',\n",
       " '19-month-old stage',\n",
       " 'mature stage',\n",
       " '80-year-old stage',\n",
       " 'ninth decade stage',\n",
       " '81-year-old stage',\n",
       " '82-year-old stage',\n",
       " '83-year-old stage',\n",
       " '84-year-old stage',\n",
       " '85-year-old stage',\n",
       " '86-year-old stage',\n",
       " '87-year-old stage',\n",
       " '88-year-old stage',\n",
       " '89-year-old stage',\n",
       " '90-year-old stage',\n",
       " 'tenth decade stage',\n",
       " '91-year-old stage',\n",
       " '92-year-old stage',\n",
       " '93-year-old stage',\n",
       " '94-year-old stage',\n",
       " '95-year-old stage',\n",
       " '96-year-old stage',\n",
       " '97-year-old stage',\n",
       " 'prime adult stage',\n",
       " 'adult stage',\n",
       " 'young adult stage',\n",
       " 'middle aged stage',\n",
       " 'child stage (1-4 yo)',\n",
       " 'under-1-year-old stage',\n",
       " 'postnatal stage',\n",
       " 'pediatric stage',\n",
       " 'nursing stage (0-11 months)']"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ontology_infos['development_stage']['labels']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "174d1c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_symb_to_gene_id = {}\n",
    "gene_synonym_to_gene_id = {}\n",
    "with open(gene_info_path, 'r') as fp:\n",
    "    reader = csv.reader(fp, delimiter='\\t')\n",
    "    next(reader)\n",
    "    for row in reader:\n",
    "        gene_symb_to_gene_id[row[2]] = row[0]\n",
    "        gene_synonym_to_gene_id[row[4]] = row[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9864a62a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41043"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_symb_to_gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "febeb3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(os.path.join(data_root, fnames[0]), sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dce0fe7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nuc-RM101-2_AAACCTGAGTCATCCA-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGCAATACGCT-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGCACGAGGTA-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGCAGATGAGC-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGCATAAGACA-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGGTGTCCTCT-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGGTGTGACGA-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGTCAGTACGT-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGTCGCCTGTT-1</th>\n",
       "      <th>Nuc-RM101-2_AAACCTGTCTGGGCCA-1</th>\n",
       "      <th>...</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCAAGTTAGGTA-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCACATCGGTTA-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCACATTGGGCC-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCAGTATGGTTC-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCAGTCGACTAT-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCAGTCGCTTCT-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCAGTCTGGTCG-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCATCACTATTC-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCATCAGTACGT-1</th>\n",
       "      <th>Nuc-RM101-2_TTTGTCATCTCTTATG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268895</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000148584</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000245105</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000070476</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000203995</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000162378</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000159840</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000074755</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21283 rows × 5491 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Nuc-RM101-2_AAACCTGAGTCATCCA-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGCAATACGCT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGCACGAGGTA-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGCAGATGAGC-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGCATAAGACA-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               1   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGGTGTCCTCT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               1   \n",
       "ENSG00000159840                               1   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGGTGTGACGA-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               1   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGTCAGTACGT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               1   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGTCGCCTGTT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_AAACCTGTCTGGGCCA-1  ...  \\\n",
       "gene_symbol                                      ...   \n",
       "ENSG00000121410                               0  ...   \n",
       "ENSG00000268895                               0  ...   \n",
       "ENSG00000148584                               0  ...   \n",
       "ENSG00000175899                               0  ...   \n",
       "ENSG00000245105                               0  ...   \n",
       "...                                         ...  ...   \n",
       "ENSG00000070476                               0  ...   \n",
       "ENSG00000203995                               0  ...   \n",
       "ENSG00000162378                               0  ...   \n",
       "ENSG00000159840                               0  ...   \n",
       "ENSG00000074755                               0  ...   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCAAGTTAGGTA-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCACATCGGTTA-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCACATTGGGCC-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               1   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCAGTATGGTTC-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCAGTCGACTAT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               1   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCAGTCGCTTCT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               1   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCAGTCTGGTCG-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               0   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCATCACTATTC-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               0   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               1   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               1   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCATCAGTACGT-1  \\\n",
       "gene_symbol                                       \n",
       "ENSG00000121410                               0   \n",
       "ENSG00000268895                               0   \n",
       "ENSG00000148584                               0   \n",
       "ENSG00000175899                               1   \n",
       "ENSG00000245105                               0   \n",
       "...                                         ...   \n",
       "ENSG00000070476                               0   \n",
       "ENSG00000203995                               0   \n",
       "ENSG00000162378                               1   \n",
       "ENSG00000159840                               0   \n",
       "ENSG00000074755                               0   \n",
       "\n",
       "                 Nuc-RM101-2_TTTGTCATCTCTTATG-1  \n",
       "gene_symbol                                      \n",
       "ENSG00000121410                               0  \n",
       "ENSG00000268895                               1  \n",
       "ENSG00000148584                               0  \n",
       "ENSG00000175899                               0  \n",
       "ENSG00000245105                               0  \n",
       "...                                         ...  \n",
       "ENSG00000070476                               0  \n",
       "ENSG00000203995                               0  \n",
       "ENSG00000162378                               1  \n",
       "ENSG00000159840                               0  \n",
       "ENSG00000074755                               0  \n",
       "\n",
       "[21283 rows x 5491 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_symbols = df.index.to_series(name='gene_symbol')\n",
    "mapped_ids = original_symbols.map(lambda s: gene_symb_to_gene_id.get(s, s))\n",
    "df.index = mapped_ids\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "322737bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>total_mrna_umis</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>sex</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_AAACCTGAGTCATCCA-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>1919</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_AAACCTGCAATACGCT-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>4253</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_AAACCTGCACGAGGTA-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>2146</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_AAACCTGCAGATGAGC-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>1752</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_AAACCTGCATAAGACA-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>1781</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_TTTGTCAGTCGCTTCT-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>2034</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_TTTGTCAGTCTGGTCG-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>1756</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_TTTGTCATCACTATTC-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>6139</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_TTTGTCATCAGTACGT-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>1419</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nuc-RM101-2_TTTGTCATCTCTTATG-1</th>\n",
       "      <td>nucleus</td>\n",
       "      <td>6273</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5491 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               suspension_type  total_mrna_umis  \\\n",
       "Nuc-RM101-2_AAACCTGAGTCATCCA-1         nucleus             1919   \n",
       "Nuc-RM101-2_AAACCTGCAATACGCT-1         nucleus             4253   \n",
       "Nuc-RM101-2_AAACCTGCACGAGGTA-1         nucleus             2146   \n",
       "Nuc-RM101-2_AAACCTGCAGATGAGC-1         nucleus             1752   \n",
       "Nuc-RM101-2_AAACCTGCATAAGACA-1         nucleus             1781   \n",
       "...                                        ...              ...   \n",
       "Nuc-RM101-2_TTTGTCAGTCGCTTCT-1         nucleus             2034   \n",
       "Nuc-RM101-2_TTTGTCAGTCTGGTCG-1         nucleus             1756   \n",
       "Nuc-RM101-2_TTTGTCATCACTATTC-1         nucleus             6139   \n",
       "Nuc-RM101-2_TTTGTCATCAGTACGT-1         nucleus             1419   \n",
       "Nuc-RM101-2_TTTGTCATCTCTTATG-1         nucleus             6273   \n",
       "\n",
       "                               assay_ontology_term_id      assay     sex  \\\n",
       "Nuc-RM101-2_AAACCTGAGTCATCCA-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_AAACCTGCAATACGCT-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_AAACCTGCACGAGGTA-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_AAACCTGCAGATGAGC-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_AAACCTGCATAAGACA-1            EFO:0009899  10x 3' v2  Female   \n",
       "...                                               ...        ...     ...   \n",
       "Nuc-RM101-2_TTTGTCAGTCGCTTCT-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_TTTGTCAGTCTGGTCG-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_TTTGTCATCACTATTC-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_TTTGTCATCAGTACGT-1            EFO:0009899  10x 3' v2  Female   \n",
       "Nuc-RM101-2_TTTGTCATCTCTTATG-1            EFO:0009899  10x 3' v2  Female   \n",
       "\n",
       "                               sex_ontology_term_id  \n",
       "Nuc-RM101-2_AAACCTGAGTCATCCA-1         PATO:0000383  \n",
       "Nuc-RM101-2_AAACCTGCAATACGCT-1         PATO:0000383  \n",
       "Nuc-RM101-2_AAACCTGCACGAGGTA-1         PATO:0000383  \n",
       "Nuc-RM101-2_AAACCTGCAGATGAGC-1         PATO:0000383  \n",
       "Nuc-RM101-2_AAACCTGCATAAGACA-1         PATO:0000383  \n",
       "...                                             ...  \n",
       "Nuc-RM101-2_TTTGTCAGTCGCTTCT-1         PATO:0000383  \n",
       "Nuc-RM101-2_TTTGTCAGTCTGGTCG-1         PATO:0000383  \n",
       "Nuc-RM101-2_TTTGTCATCACTATTC-1         PATO:0000383  \n",
       "Nuc-RM101-2_TTTGTCATCAGTACGT-1         PATO:0000383  \n",
       "Nuc-RM101-2_TTTGTCATCTCTTATG-1         PATO:0000383  \n",
       "\n",
       "[5491 rows x 6 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "    'suspension_type': ['nucleus'] * len(df.columns),\n",
    "    'total_mrna_umis': df.sum(axis=0),\n",
    "    'assay_ontology_term_id': ['EFO:0009899'] * len(df.columns),\n",
    "    'assay': [\"10x 3' v2\"] * len(df.columns),\n",
    "    'sex': [\"Female\"] * len(df.columns),\n",
    "    'sex_ontology_term_id': [\"PATO:0000383\"] * len(df.columns)\n",
    "}\n",
    "\n",
    "obs = pd.DataFrame(index=df.columns, data=data)\n",
    "obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "061c8c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs = pd.DataFrame(index=df.columns)      # one row per sample\n",
    "var = pd.DataFrame(index=df.index)        # one row per gene ID\n",
    "var['gene_symbol'] = original_symbols     # store the original symbol\n",
    "\n",
    "adata = AnnData(X=df.values.T, obs=obs, var=var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9fb726e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000121410', 'ENSG00000268895', 'ENSG00000148584',\n",
       "       'ENSG00000175899', 'ENSG00000245105', 'ENSG00000166535',\n",
       "       'ENSG00000256661', 'ENSG00000184389', 'ENSG00000128274',\n",
       "       'ENSG00000094914',\n",
       "       ...\n",
       "       'ENSG00000086827', 'ENSG00000174442', 'ENSG00000122952',\n",
       "       'ENSG00000198205', 'ENSG00000198455', 'ENSG00000070476',\n",
       "       'ENSG00000203995', 'ENSG00000162378', 'ENSG00000159840',\n",
       "       'ENSG00000074755'],\n",
       "      dtype='object', name='gene_symbol', length=21283)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "978f96b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21-Apr-2025 00:58:27 DEBUG utils - Directory . already exists. Skipping.\n",
      "21-Apr-2025 00:58:27 INFO GEOparse - File already exist: using local version.\n",
      "21-Apr-2025 00:58:27 INFO GEOparse - Parsing ./GSE153807_family.soft.gz: \n",
      "21-Apr-2025 00:58:27 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "21-Apr-2025 00:58:27 DEBUG GEOparse - SERIES: GSE153807\n",
      "21-Apr-2025 00:58:27 DEBUG GEOparse - PLATFORM: GPL20301\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654466\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654467\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654468\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654469\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654470\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654471\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654472\n",
      "21-Apr-2025 00:58:28 DEBUG GEOparse - SAMPLE: GSM4654473\n"
     ]
    }
   ],
   "source": [
    "# Download the GSE and its supplementary files\n",
    "# gse = GEOparse.get_GEO(geo=\"GSE153807\", destdir=\".\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "793b193d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SERIES: GSE153807 - 8 SAMPLES, 1 d(s)>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "55121de3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM example:\n",
      "Name:  GSM4654466\n",
      "Metadata:\n",
      " - title : Nuc-RM101-1\n",
      " - geo_accession : GSM4654466\n",
      " - status : Public on Sep 29 2020\n",
      " - submission_date : Jul 04 2020\n",
      " - last_update_date : Sep 29 2020\n",
      " - type : SRA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : Temporal cortex\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - characteristics_ch1 : age: 50, Sex: Female, sequencing_date: 201908, sample: S1, patient: RM101\n",
      " - molecule_ch1 : polyA RNA\n",
      " - extract_protocol_ch1 : Brain tissue was sliced on dry ice, then homogenized manually (15 gentle strokes) with 1mL ice-cold  Homogenisation Buffer (HB) with 5µL Rnasin Plus. The homogenate was strained with a 70µm strainer and washed with 1.65mL to a final volume of 2.65mL. 2.65mL of Gradient medium was added (Vf = 5.3mL). To isolate the nuclei, the sample was added to a 4mL 29% cushion using a P1000, and the weight adjusted with HB. The sample was centrifuged in a SW41Ti rotor at 7,700 rpm for 30 minutes at 4°C. The supernatant was removed with a plastic Pasteur pipette, followed by removal of the lower supernatant with P200. Nuclei were resuspended in 200µL of resuspension buffer, transferred to a new tube, washed again with 100-200µL resuspension buffer, and pooled with the previous solution. Clumps were disrupted by pipetting with P200, then filtered through a Falcon tube with 0.35µm strainer. 9µL of sample was mixed with 1µL of propidium iodide (PI) stain, loaded onto a LUNA-FL slide and allowed to settle for 30 seconds. We viewed nuclei with the LUNA-FL Automated cell counter to check numbers and shape., RNA sequencing was performed using the 10x Genomics Single Cell 3` Reagent Kit (v2) according to manufacturer protocols. cDNA libraries from fresh-frozen nuclei were sequenced on an Illumina HiSeq platform 4000.\n",
      " - description : RM101, Single nucleus sequencing\n",
      " - data_processing : Alignment. Cellranger v2.1.1 was used to demultiplex and align sequencing output to a human reference genome. We used a “pre-mRNA” database to align single nuclei to exons as well as introns. Following alignment, nuclei from one patient sample (RM101.1) were removed due to poor quality (low read and gene count)., Extraction of microglial nuclei. Data was processed using the Seurat v3.0.2 package (Butler et al., 2018; Stuart et al., 2019) in R v3.6.1. For each patient, the count matrix was filtered to exclude nuclei with fewer than 100 genes. Counts were normalised for library size, scaled by 10,000 and log-transformed using the natural log. FindVariableFeatures was run using a variance-stabilising transformation (“vst”) to identify the 2,000 most variable genes in each sample., Data from the 4 patients was then integrated using Seurat’s FindIntegrationAnchors with default parameters, and IntegrateData using 40 principal components (PCs). The dataset consisted of 37,060 nuclei, with a mean read count of 4,305 counts per nucleus, and 1,791 genes per nucleus. Integrated data was scaled (default Seurat parameters). We ran a Principal Components Analysis (PCA), then calculated Uniform Manifold Approximation and Projection (UMAP) embeddings using 40 PCs. We identified clusters using Seurat’s FindNeighbours and FindClusters functions, again using 40 PCs., Microglial clusters were identified using known markers including P2RY12, CSF1R, CX3CR1, and extracted for downstream analysis., Pre-processing of microglial nuclei per patient. Microglia from each patient sample were analysed individually as described for all cell types above, with the following modifications: raw counts were filtered to remove genes and counts that were ± 3 standard deviations away from the median value. After normalization, doublets were identified using DoubletFinder v2.0.2 (McGinnis et al., 2019) using 40 PCs, assuming a 7.5% doublet rate. Following removal of doublets, filtering and Seurat normalization were performed again. Data from patients was then integrated and clusters were identified as above. We discarded small clusters than contained markers for microglia as well as other cell types., Genome_build: hg38 build 95, Supplementary_files_format_and_content: count matrices for each sample\n",
      " - platform_id : GPL20301\n",
      " - contact_name : Bart,,de Strooper\n",
      " - contact_email : bart.destrooper@kuleuven.be\n",
      " - contact_department : VIB-KU Leuven Center for Brain & Disease Research\n",
      " - contact_institute : KULeuven\n",
      " - contact_address : Campus Gasthuisberg, Herestraat 49, bus 602\n",
      " - contact_city : Leuven\n",
      " - contact_zip/postal_code : 3000\n",
      " - contact_country : Belgium\n",
      " - instrument_model : Illumina HiSeq 4000\n",
      " - library_selection : cDNA\n",
      " - library_source : transcriptomic\n",
      " - library_strategy : RNA-Seq\n",
      " - relation : BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN15448195, SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX8667031\n",
      " - supplementary_file_1 : NONE\n",
      " - series_id : GSE153807\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "<bound method NDFrame.head of Empty DataFrame\n",
      "Columns: []\n",
      "Index: []>\n",
      "\n",
      "GSM:\n",
      "<SAMPLE: GSM4654466>\n"
     ]
    }
   ],
   "source": [
    "# print(\"GSM example:\")\n",
    "# for gsm_name, gsm in gse.gsms.items():\n",
    "#     print(\"Name: \", gsm_name)\n",
    "#     print(\"Metadata:\",)\n",
    "#     for key, value in gsm.metadata.items():\n",
    "#         print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "#     print (\"Table data:\",)\n",
    "#     print (gsm.table.head)\n",
    "#     print()\n",
    "#     print(\"GSM:\")\n",
    "#     print(gsm)\n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e65a796e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GPL example:\n",
      "Name:  GPL20301\n",
      "Metadata:\n",
      " - title : Illumina HiSeq 4000 (Homo sapiens)\n",
      " - geo_accession : GPL20301\n",
      " - status : Public on Jun 09 2015\n",
      " - submission_date : Jun 09 2015\n",
      " - last_update_date : Mar 21 2019\n",
      " - technology : high-throughput sequencing\n",
      " - distribution : virtual\n",
      " - organism : Homo sapiens\n",
      " - taxid : 9606\n",
      " - contact_name : ,,GEO\n",
      " - contact_country : USA\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# print()\n",
    "# print(\"GPL example:\")\n",
    "# for gpl_name, gpl in gse.gpls.items():\n",
    "#     print(\"Name: \", gpl_name)\n",
    "#     print(\"Metadata:\",)\n",
    "#     for key, value in gpl.metadata.items():\n",
    "#         print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "#     print(\"Table data:\",)\n",
    "#     print(gpl.table.head())\n",
    "#     break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
