{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "946a96f5",
   "metadata": {},
   "source": [
    "<center><h1> Benchmarking Datasets </h1></center>\n",
    "\n",
    "## Context\n",
    "\n",
    "This notebook intend to load the dataset used in GEARS and in the benchmarking paper [\\[1\\]](#benchmark).\n",
    "\n",
    "In particular, we will be using the following [datasets](#datasets).\n",
    "\n",
    "## Table of Contents\n",
    "\n",
    "- [Raw Data](#raw-pipeline)\n",
    "- [Pre-processed Data](#pre-processed-pipeline)\n",
    "\n",
    "## Datasets\n",
    "\n",
    "### Raw Data\n",
    "\n",
    "- [Adamson](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90546): \n",
    "  - ht<span>tps://</span>www.ncbi.nlm.nih.gov/geo/download/?acc=GSE90546&format=file\n",
    "  - Assay type: perturb-seq: a combination of droplet based scRNAseq with a strategy for barcoding CRISPR-mediated perturbations.\n",
    "  - Single CRIPSRi perturbation.\n",
    "  - Overall: 3 different pooled CRIPSR screening experiments were conducted via perturb-seq.\n",
    "- [Replogle](https://plus.figshare.com/articles/dataset/_Mapping_information-rich_genotype-phenotype_landscapes_with_genome-scale_Perturb-seq_Replogle_et_al_2022_processed_Perturb-seq_datasets/20029387):\n",
    "  - Single cell raw **K562 genome scale scale day 8 post-transduction** ht<span>tps://</span>plus.figshare.com/ndownloader/files/35775507\n",
    "  - Single cell raw **K562 essential scale day 6 post-transduction** ht<span>tps://</span>plus.figshare.com/ndownloader/files/35773219\n",
    "  - Single cell raw **RPE1 essential scale  day 7 post-transduction** ht<span>tps://</span>plus.figshare.com/ndownloader/files/35775606\n",
    "- [Norman](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE133344):\n",
    "  - raw_barcodes.tsv.gz: ht<span>tps://</span>ftp.ncbi.nlm.nih.gov/geo/series/GSE133nnn/GSE133344/suppl/GSE133344%5Fraw%5Fbarcodes.tsv.gz\n",
    "  - raw_cell_identities.csv.gz: ht<span>tps://</span>ftp.ncbi.nlm.nih.gov/geo/series/GSE133nnn/GSE133344/suppl/GSE133344%5Fraw%5Fcell%5Fidentities.csv.gz\n",
    "  - raw_genes.tsv.gz: ht<span>tps://</span>ftp.ncbi.nlm.nih.gov/geo/series/GSE133nnn/GSE133344/suppl/GSE133344%5Fraw%5Fgenes.tsv.gz\n",
    "  - raw_matrix.mtx.gz: ht<span>tps://</span>ftp.ncbi.nlm.nih.gov/geo/series/GSE133nnn/GSE133344/suppl/GSE133344%5Fraw%5Fmatrix.mtx.gz\n",
    "\n",
    "### Pre-processed Data\n",
    "\n",
    "For completeness, here are the datasets used by GEARS or scFoundation and the benchmarking paper [\\[1\\]](#benchmark).\n",
    "\n",
    "**GEARS**:\n",
    "\n",
    "- Single-gene perturbation:\n",
    "  - [Adamson](https://dataverse.harvard.edu/file.xhtml?fileId=6154417&version=3.0&toolType=PREVIEW): ht<span>tps://</span>dataverse.harvard.edu/api/access/datafile/6154417\n",
    "  - [Relogle K562](https://dataverse.harvard.edu/file.xhtml?fileId=7458695&version=6.0&toolType=PREVIEW): ht<span>tps://</span>dataverse.harvard.edu/api/access/datafile/7458695\n",
    "  - [Replogle RPE1](https://dataverse.harvard.edu/file.xhtml?fileId=7458694&version=6.0&toolType=PREVIEW): ht<span>tps://</span>dataverse.harvard.edu/api/access/datafile/7458694\n",
    "\n",
    "- Multiple-gene perturbation\n",
    "  - [Norman](https://dataverse.harvard.edu/file.xhtml?fileId=6154020&version=3.0&toolType=PREVIEW): ht<span>tps://</span>dataverse.harvard.edu/api/access/datafile/6154020\n",
    "  - 131 two-gene perturbations\n",
    "\n",
    "**scFoundation**:\n",
    "\n",
    "Another preprocessing strategy for the Norman et al. dataset\n",
    "- [Norman scFoundation](https://figshare.com/articles/dataset/scFoundation_Large_Scale_Foundation_Model_on_Single-cell_Transcriptomics_-_processed_datasets/24049200?file=44477939): ht<span>tps://</span>figshare.com/ndownloader/files/44477939\n",
    "## References\n",
    "\n",
    "1. Ahlmann-Eltze, Huber, and Anders, “Deep Learning-Based Predictions of Gene Perturbation Effects Do Not yet Outperform Simple Linear Baselines.”<a id=\"benchmark\"></a>\n",
    "2. Adamson et al., “A Multiplexed Single-Cell CRISPR Screening Platform Enables Systematic Dissection of the Unfolded Protein Response.”<a id=\"adamson\"></a>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70975cfa",
   "metadata": {},
   "source": [
    "## 1) Download Datasets\n",
    "\n",
    "Note that the bash script is run from the Notebook but you can also run it from the shell. File will be downloaded using the `yaml` config in `fine tune/config/download.yaml`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a7b63e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../../ # cd to root of the project\n",
    "python fine_tune/scripts/download/cli.py --config fine_tune/datasets/config/benchmarking.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ead96486",
   "metadata": {},
   "source": [
    "## 2) Adamson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00b791b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All imports here\n",
    "from functools import reduce\n",
    "from pathlib import Path\n",
    "\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pooch\n",
    "from rdkit import Chem\n",
    "\n",
    "# Define compound metadata\n",
    "from rdkit.Chem import Descriptors\n",
    "from scipy.io import mmread\n",
    "\n",
    "from fine_tune.scripts import format_obs, format_var\n",
    "\n",
    "data_path = Path(\"./../datasets\")\n",
    "\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e60d694",
   "metadata": {},
   "source": [
    "**Note on Adamson**\n",
    "\n",
    "| Accession              | Experiment                     | Model                               | Protocol                                                                                    | Additional Protocol                                                                                                                                                                                                                                                                                                                                               | sgRNA Count                                                                               | Cell Count | Assay Type           |\n",
    "|------------------------|--------------------------------|-------------------------------------|---------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------|------------|----------------------|\n",
    "| **GSM2406675\\_10X001** | **Pilot Experiment**           | K562 cells with dCas9-KRAB (cBA010) | Individual transduction → pooled after 3 days → selection + 5 days combined growth          | Growth in the presence of puromycin (3µg/mL)                                                                                                                                                                                                                                                                                                                      | 8 distinct GBCs (1 control)                                                               | 5,768      | 10x 3' v1            |\n",
    "| **GSM2406677\\_10X005** | **UPR Epistasis Experiment**   | K562 cells with dCas9-KRAB (cBA010) | Individual transduction → pooled after 3 days → selection + 2 days combined growth          | To limit heterogenous effects of cell microenvironments caused by cell settling, the sorted cells were grown with continuous agitation on an orbital shaker. <br> <br> Pharmacological treatment: <br> • Thapsigargin (ER Ca²⁺ pump inhibitor) (100 nM/mL for 4 hr.)<br> • Tunicamycin (N-glycosylation inhibitor) (4 μg/mL for 6h) <br> • DMSO (control) for 6hr | 7 GBCs (3-guide vectors of 3 genes — single, double, triple) + 2 triple negative controls | 15,006     | 10x 3' v1            |\n",
    "| **GSM2406681\\_10X010** | **UPR Perturb-seq Experiment** | K562 cells with dCas9-KRAB (cBA011) | Individual lentivirus production → pooled transduction → selection + 7 days combined growth | To limit heterogenous effects of cell microenvironments caused by cell settling, the sorted cells were grown with continuous agitation on an orbital shaker. <br><br> Sequencing across two separate runs totaling 10 lanes.                                                                                                                                      | 93 GBCs (2 control GBCs; targets cover 82 genes)                                          | 65,337     | 10x 3' v1 (10 lanes) |\n",
    "\n",
    "**Remark:**\n",
    "- UPR (unfolded protein response)\n",
    "- The assay technology is assumed to be the v1 of the Chromium Single Cell 3′ Solution (No specification of the version but 2016, launched year of the technology).\n",
    "- When the identity of a Cas9 targeting single guide RNA (sgRNA) couldn't be uniquely identified, it was either:\n",
    "  - \"multiplet\" \n",
    "  - \"NaN\"\n",
    "- In the \"UPR Epistasis Experiment\" [\\[2\\]](#adamson): \n",
    "  > Note in the raw sequencing data the tunicamycin-treated cells have gemgroup 1 (as a BAM tag), the thapsigargin-treated cells have gemgroup 2, and the DMSO-treated cells have gemgroup 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e5bd494",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GSM2406675_10X001': AnnData object with n_obs × n_vars = 5768 × 35635\n",
       "     obs: 'guide identity', 'read count', 'UMI count', 'coverage', 'good coverage', 'number of cells', 'experiment',\n",
       " 'GSM2406677_10X005': AnnData object with n_obs × n_vars = 15006 × 32738\n",
       "     obs: 'guide identity', 'read count', 'UMI count', 'coverage', 'good coverage', 'number of cells', 'experiment',\n",
       " 'GSM2406681_10X010': AnnData object with n_obs × n_vars = 65337 × 32738\n",
       "     obs: 'guide identity', 'read count', 'UMI count', 'coverage', 'good coverage', 'number of cells', 'experiment'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adamson_dir = data_path.joinpath(\"raw/raw_adamson\")\n",
    "mtx_files = sorted(adamson_dir.glob(\"*_matrix.mtx.txt.gz\"))\n",
    "\n",
    "all_adatas = {}\n",
    "\n",
    "for mtx_file in mtx_files:\n",
    "    prefix = mtx_file.name.split(\"_matrix\")[0]\n",
    "\n",
    "    # Paths\n",
    "    barcodes_file = adamson_dir / f\"{prefix}_barcodes.tsv.gz\"\n",
    "    genes_file = adamson_dir / f\"{prefix}_genes.tsv.gz\"\n",
    "    identities_file = adamson_dir / f\"{prefix}_cell_identities.csv.gz\"\n",
    "\n",
    "    # Load metadata\n",
    "    genes = pd.read_csv(genes_file, header=None, sep=\"\\t\")\n",
    "    barcodes = pd.read_csv(barcodes_file, header=None)\n",
    "    cell_identities = pd.read_csv(identities_file, index_col=0)\n",
    "\n",
    "    # Create AnnData\n",
    "    adata = ad.io.read_mtx(mtx_file).T\n",
    "    adata.var_names = genes[0].values\n",
    "    adata.obs_names = barcodes[0].values\n",
    "    # adata.obs = cell_identities.loc[adata.obs_names] fail because more barcodes than cell_identity.\n",
    "    adata.obs = pd.DataFrame(index=adata.obs_names)\n",
    "    adata.obs = adata.obs.join(cell_identities) # fill with NaN barcodes with non found identity\n",
    "\n",
    "    # Optional: keep track of the experiment\n",
    "    adata.obs[\"experiment\"] = prefix\n",
    "    all_adatas[prefix] = adata\n",
    "\n",
    "all_adatas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "befc5aa0",
   "metadata": {},
   "source": [
    "### a) Retrieve correct perturbation\n",
    "\n",
    "Let's check the number of unique guide barcode which shall corresponds to the unique perturbations (except than certain barcode targets the same genes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "acab6109",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique 'guide barcode' detected:\n",
      "GSM2406675_10X001: 10, expected 8, of which 1 controls. \n",
      "GSM2406677_10X005: 21, expected 9, of which 2 controls. \n",
      "GSM2406681_10X010: 115, expected 93, of which 2 controls. \n",
      "Potentially NaN or '*' for non identifiable or multiplets should be removed.\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    \"Number of unique 'guide barcode' detected:\\n\" +\n",
    "    \"\\n\".join([\n",
    "        f\"{exp}: {adata.obs['guide identity'].unique().__len__()}, \"\n",
    "        f\"expected {gbc}, of which {control} controls. \"\n",
    "        for (exp, adata), (gbc, control) in zip(\n",
    "            all_adatas.items(),\n",
    "            zip([8, 9, 93],[1, 2, 2]) # issued from the table above, sgRNA column\n",
    "        )]) +\n",
    "    \"\\nPotentially NaN or '*' for non identifiable or multiplets should be removed.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1db2fd57",
   "metadata": {},
   "source": [
    "**Filtering must occurs:**\n",
    "- GSM2406675_10X001: \n",
    "  - Remove Nan and '*'\n",
    "- GSM2406677_10X005: \n",
    "  - Remove Nan and '*'\n",
    "  - Targeted genes should be a single, double or triple combination of:\n",
    "    - ATF6\n",
    "    - IRE1\n",
    "    - PERK\n",
    "- GSM2406681_10X010: \n",
    "  - Remove Nan and '*'\n",
    "  - Shall use the \"Table S1. Protospacer Sequences of sgRNAs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2e2cfe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adamson_pilot = all_adatas[\"GSM2406675_10X001\"]\n",
    "adamson_epistasis = all_adatas[\"GSM2406677_10X005\"]\n",
    "adamson_perturb = all_adatas[\"GSM2406681_10X010\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "abad69e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2294375/2038629506.py:3: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  ~adamson_pilot.obs[\"guide identity\"].astype(str).str.contains(r\"(nan|\\*)\")\n",
      "/tmp/ipykernel_2294375/2038629506.py:7: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  adamson_epistasis.obs[\"guide identity\"].astype(str).str.contains(\"(ctrl|IRE1|PERK|ATF6)\")\n",
      "/home/hhakem/projects/cellarium-ml/.venv/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "# Pilot experiment - remove NaN and '*'\n",
    "adamson_pilot = adamson_pilot[\n",
    "    ~adamson_pilot.obs[\"guide identity\"].astype(str).str.contains(r\"(nan|\\*)\")\n",
    "]\n",
    "# UPR epistasis experiment - keep only guide containing any of ctrl, IRE1, PERK, ATF6\n",
    "adamson_epistasis = adamson_epistasis[\n",
    "    adamson_epistasis.obs[\"guide identity\"].astype(str).str.contains(\"(ctrl|IRE1|PERK|ATF6)\")\n",
    "]\n",
    "\n",
    "# UPR perturb seq experiment - retrieve table S1 and filter out guide while keeping control: 'mod'\n",
    "file = pooch.retrieve(\n",
    "    url=\"https://ars.els-cdn.com/content/image/1-s2.0-S0092867416316609-mmc1.xlsx\",\n",
    "    known_hash=\"sha256:9b5935cb15ba2f6d60d3017832de2918e7d4f172db6f202be7999cba5feea82b\"\n",
    ")\n",
    "protospace_df = pd.read_excel(file, header=1)\n",
    "# Some gene names are separated with '/' for aliases.\n",
    "# Since the convention chosen in the dataset is not known, corresponding rows are demultiplicated\n",
    "protospace_df[\"Gene\"] = protospace_df[\"Gene\"].apply(lambda x: x.split(\"/\") if isinstance(x, str) else x)\n",
    "protospace_df = protospace_df.explode(\"Gene\").reset_index(drop=True)\n",
    "\n",
    "adamson_perturb = adamson_perturb[(\n",
    "    # retrieve guide identity in the protospace table\n",
    "    (pert := adamson_perturb.obs[\"guide identity\"]).isin(\n",
    "        (protospace_df[\"Gene\"].dropna() + \"_\" + protospace_df[\"Perturb-seq_Vector_ID\"]).dropna()) |\n",
    "    # retrieve control with large occurence (more than 100 cells)\n",
    "    (pert.str.contains(\"mod\") & pert.isin(pert.value_counts()[pert.value_counts() > 100].index))\n",
    ")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc22c708",
   "metadata": {},
   "source": [
    "### b) Formatting Adamson\n",
    "\n",
    "Add more metadata:\n",
    "- `obs`\n",
    "  - `gene_perturbed`\n",
    "    - parse `guide identity`:\n",
    "      - **Renaming target:**\n",
    "        - GSM2406675_10X001\n",
    "          - Control is written with 'mod'. Should be renamed in 'NT' (non targeting)\n",
    "        - GSM2406677_10X005\n",
    "          - IRE1 -> ERN1 to match the actual gene name\n",
    "          - PERK -> EIF2AK3 to match the actual gene name\n",
    "          - Single perturbation should have the 'only' removed and add NT+NT since triple guides are used\n",
    "          - Double or Triple perturbations should be spaced with '+'\n",
    "          - Controls are written with 'ctrl'. Since it is a triple guide, should be marked as NT+NT+NT\n",
    "        - GSM2406681_10X010\n",
    "          - Controls are written with 'ctrl'. Should be renamed in 'NT'\n",
    "  - `tissue` \n",
    "    - cell line\n",
    "  - `cell_type`\n",
    "    - lymphoblasts\n",
    "  - `cell_line` \n",
    "    - K562\n",
    "  - `disease` \n",
    "    - chronic myelogenous leukemia\n",
    "  - `organism` \n",
    "    - Homo sapiens\n",
    "  - `assay` \n",
    "    - 10x 3' v1\n",
    "  - `cell_barcode`\n",
    "    - Reindex the obs table\n",
    "  - `perturbation_type` \n",
    "    - CRISPRi, CRISPRa, CRISPRko, drug (sum of those)\n",
    "  - `drug` \n",
    "    - (name of the drug or sum of those, (control = DMSO))\n",
    "  - `drug_canonical_smiles` \n",
    "    - SMILES representation of the drug. Should be canonicalized.\n",
    "  - `drug_dose` \n",
    "    - (dosage of the drug in µM). Should be formatted as a list of the dose: `[(drug1, durg_dose1, drug_unit1), ...]`\n",
    "  - `drug_targets` \n",
    "    - (name of the drug targeted gene, NaN if unknown) Different targets are separated by a `\",\"`\n",
    "  - `drug_moa` \n",
    "    - (name of the drug mechanism of action, NaN if unknown). Different MoA are separated by a `\",\"`\n",
    "  - `gene_perturbed` \n",
    "    - (name of the gene perturbed or sum of those (control = NT) (if single perturbation but double or triple guide, should be gene+NT / gene+NT+NT))\n",
    "  - `perturbation_id` \n",
    "    - **Assigning unique perturbation identifiers**\n",
    "  \n",
    "      To facilitate downstream analysis, we add a `perturbation_id` column that uniquely identifies the biological perturbation applied in each sample. This string encodes key information from the metadata, such as the perturbation type (e.g., CRISPRi, drug), the target gene or compound, and the dose when applicable.\n",
    "\n",
    "      This is useful for:\n",
    "      - Grouping cells by perturbation\n",
    "      - Comparing effects across perturbations\n",
    "      - Ensuring consistent identifiers across datasets\n",
    "\n",
    "      For example, a row with:\n",
    "      - `perturbation_type = \"CRISPRi\"`\n",
    "      - `gene_perturbed = \"TP53\"`\n",
    "\n",
    "      Will yield:  \n",
    "      `perturbation_id = \"TP53_CRISPRi\"`\n",
    "\n",
    "      A drug-based example:\n",
    "      - `perturbation_type = \"drug\"`\n",
    "      - `drug = \"bortezomib\"`\n",
    "      - `drug_dose = [(\"bortezomib\", 0.1, \"uM\")]`\n",
    "\n",
    "      Will yield:  \n",
    "      `perturbation_id = \"bortezomib_0.1(uM)\"`\n",
    "  - `is_control` \n",
    "    - True if drug = DMSO and gene_perturbed = NaN, or if gene_perturbed.split(\"+\") contains only NT. else False.\n",
    "\n",
    "- `var`\n",
    "  - `gene_name`\n",
    "    - Join with `genome.py`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa4c7e09",
   "metadata": {},
   "source": [
    "#### i) obs\n",
    "\n",
    "**Rename control and Parse guide identity**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b4853dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2294375/2919193891.py:6: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  ds.obs[\"gene_perturbed\"] = ds.obs[\"guide identity\"].where(\n",
      "/tmp/ipykernel_2294375/2919193891.py:6: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  ds.obs[\"gene_perturbed\"] = ds.obs[\"guide identity\"].where(\n",
      "/tmp/ipykernel_2294375/2919193891.py:6: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  ds.obs[\"gene_perturbed\"] = ds.obs[\"guide identity\"].where(\n"
     ]
    }
   ],
   "source": [
    "for ds, pattern, replacement in [\n",
    "    (adamson_pilot, r\"\\(mod\\)\", \"NT\"),\n",
    "    (adamson_epistasis, \"ctrl\", \"NT+NT+NT\"), # those are triple guide control\n",
    "    (adamson_perturb, r\"\\(mod\\)\", \"NT\"),\n",
    "]:\n",
    "    ds.obs[\"gene_perturbed\"] = ds.obs[\"guide identity\"].where(\n",
    "        ~ds.obs[\"guide identity\"].str.contains(pattern), replacement\n",
    "    )\n",
    "\n",
    "# simply remove the guide id\n",
    "adamson_pilot.obs[\"gene_perturbed\"] = adamson_pilot.obs[\"gene_perturbed\"].str.split(\"_\").apply(lambda x: x[0])\n",
    "\n",
    "# remove guide id,\n",
    "# change \"IRE1\" -> \"ERN1\" and \"PERK\" -> \"EIF2AK3\"\n",
    "# make sure simple or double perturbation remains marked\n",
    "# as +NT+NT or +NT because triple guide were used for that experiment\n",
    "adamson_epistasis.obs[\"gene_perturbed\"] = adamson_epistasis.obs[\"gene_perturbed\"].str.split(\"_\").apply(\n",
    "    lambda row: row[0] if len(row) == 1 else (\n",
    "        \"+\".join(\n",
    "            [\n",
    "                {\"only\": \"NT\", \"IRE1\": \"ERN1\", \"PERK\": \"EIF2AK3\"}.get(x, x)\n",
    "                for x in row[:-1]\n",
    "            ] + ([\"NT\"] if len(row[:-1]) < 3 else [])\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "# simply remove the guide id\n",
    "adamson_perturb.obs[\"gene_perturbed\"] = adamson_perturb.obs[\"gene_perturbed\"].str.split(\"_\").apply(lambda x: x[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f5629b1",
   "metadata": {},
   "source": [
    "**Add chemical treatment for the UPR epistasis experiment**\n",
    "\n",
    "We use the BAM tag number at the end of each barcode to identify the chemical treatment, cf [\\[2\\]](#adamson)\n",
    "\n",
    "- BAM-1 tunicamycin\n",
    "  - Paper Dosage: 4µg/mL 6hr\n",
    "  - SMILES: \n",
    "    ```\n",
    "    CC(C)CCCCCCCC\\C=C\\C(=O)N[C@@H]1[C@@H](O)[C@@H](O)[C@@H](C[C@@H](O)[C@H]2O[C@H]([C@H](O)[C@@H]2O)N3C=CC(=O)NC3=O)O[C@H]1O[C@@H]4O[C@@H](CO)[C@H](O)[C@@H](O)[C@@H]4NC(C)=O\n",
    "    ```\n",
    "    - Source: [tunicamycin (Sigma, T7765)](https://www.sigmaaldrich.com/US/en/product/sigma/t7765) (Assumed same provider as Thapsigargin (Sigma)).\n",
    "    - There is multiple version, notably the fatty carbon chain has a length from 8 to 11. Here (CH2)8 is chosen, and it matches the SMILES given in the Connectivity Map.\n",
    "  \n",
    "  Other info fetched on: [Thapsigargin Connectivity Map Broad Institute](https://clue.io/command?q=thapsigargin)\n",
    "  - MoA:\n",
    "    - GLCNAC phosphotransferase inhibitor\n",
    "  - Target:\n",
    "    - DPAGT1, GNPTAB\n",
    "    - Remark: DPAGT1 is known to be targeted by tunicamycin, (source: [DPAGT1 gene card](https://www.genecards.org/cgi-bin/carddisp.pl?gene=DPAGT1)). GNPTAB is solely reported by the connectivity Map but is not necessarily chemically tested.\n",
    "  - ID:\n",
    "    - Broad: BRD-K10573841\n",
    "    - PubChem: 16220051\n",
    "\n",
    "- BAM-2 thapsigargin (Sigma, T9033),\n",
    "  - Paper Dosage: 100 nM 4hr\n",
    "  - SMILES:\n",
    "    ```\n",
    "    [H][C@@]12C([C@H](OC([C@]3(O)C)=O)[C@]3(O)[C@@H](OC(CCC)=O)C[C@]2(C)OC(C)=O)=C(C)[C@H](OC(/C(C)=C(C)/[H])=O)[C@H]1OC(CCCCCCC)=O\n",
    "    ```\n",
    "    - Source: [thapsigargin (Sigma, T9033)](https://www.sigmaaldrich.com/US/en/product/sigma/t9033)\n",
    "    - The SMILES provided from the Connectivity Map omit the stereochemistry, yet it is provided in the SIGMA product info, so it is used instead.\n",
    "  \n",
    "  Other info fetched on: [Thapsigargin Connectivity Map Broad Institute](https://clue.io/command?q=thapsigargin)\n",
    "  - MoA:\n",
    "    - ATPase inhibitor\n",
    "  - Target: \n",
    "    - ATP2A1, ATP2A2, ATP2A3\n",
    "    - Remark: ATP2A1 has the strongest binding affinity  (Kd < 100 nM) (source: [thapsigargin LINCS HMS Database](https://web.archive.org/web/20250425124838/https://lincs.hms.harvard.edu/db/sm/10293-999/)) and it is also reported in the Connectivity Map. ATP2A2 and ATP2A3 are reported with moderate binding affinity:  (100 nM < Kd < 1µM)\n",
    "  - ID\n",
    "    - Broad: BRD-A62809825\n",
    "    - PubChem: 5353993\n",
    "  \n",
    "\n",
    "- BAM-3 DMSO\n",
    "\n",
    "Additionally, `is_control = True`, should now be updated so that only DMSO chemically treated + no gene perturbation are said to be control. \n",
    "\n",
    "**Note:**\n",
    "\n",
    "`format_obs` will take care of formatting:\n",
    "- `drug_canonical_smiles`\n",
    "  - As long as it is a SMILES, will be canonicalize with `rdkit`\n",
    "- `drug_dose`\n",
    "  - As long as it is a `int`, `list[int]` or `list[tuple(str,int)]` for multiple dose / dose and unit.\n",
    "  - As a rule of thumb, the unit is always given in `uM`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c4623b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_molar_mass(smiles: str) -> float:\n",
    "    \"\"\"\n",
    "    Calculate the molar mass (molecular weight) of a molecule from its SMILES string.\n",
    "\n",
    "    Args:\n",
    "        smiles (str): The SMILES representation of the molecule.\n",
    "\n",
    "    Returns:\n",
    "        float: The molecular weight in g/mol.\n",
    "    \"\"\"\n",
    "    mol = Chem.MolFromSmiles(smiles)\n",
    "    return Descriptors.MolWt(mol)  # in g/mol\n",
    "\n",
    "\n",
    "tunicamycin_smiles = (\n",
    "    \"CC(C)CCCCCCCC\\\\C=C\\\\C(=O)N[C@@H]1[C@@H](O)[C@@H](O)[C@@H](C[C@@H](O)[C@H]2O[C@H]([C@H](O)[C@@H]2O)\"\n",
    "    \"N3C=CC(=O)NC3=O)O[C@H]1O[C@@H]4O[C@@H](CO)[C@H](O)[C@@H](O)[C@@H]4NC(C)=O\"\n",
    ")\n",
    "# 4 µg/mL → g/L -> mol/L -> umol/L = uM (micro molar)\n",
    "tunicamycin_dose = np.round((4 * 1e-6 * 1e3 / get_molar_mass(tunicamycin_smiles)) * 1e6, decimals=1)\n",
    "\n",
    "thapsigargin_smiles = (\n",
    "    \"[H][C@@]12C([C@H](OC([C@]3(O)C)=O)[C@]3(O)[C@@H](OC(CCC)=O)C[C@]2(C)OC(C)=O)=C(C)\"\n",
    "    \"[C@H](OC(/C(C)=C(C)/[H])=O)[C@H]1OC(CCCCCCC)=O\"\n",
    ")\n",
    "dmso_smiles = \"CS(=O)C\"\n",
    "\n",
    "compound_metadata = {\n",
    "    \"1\": {\n",
    "        \"drug\": \"Tunicamycin\",\n",
    "        \"drug_dose\": tunicamycin_dose,\n",
    "        \"drug_canonical_smiles\": tunicamycin_smiles,\n",
    "        \"drug_moa_broad\": \"inhibitor/antagonist\",\n",
    "        \"drug_moa_fine\": \"GLCNAC phosphotransferase inhibitor\",\n",
    "        \"drug_targets\": \"DPAGT1, GNPTAB\",\n",
    "        \"drug_pubchem_cid\": \"16220051\"\n",
    "    },\n",
    "    \"2\": {\n",
    "        \"drug\": \"Thapsigargin\",\n",
    "        \"drug_dose\": 0.1,\n",
    "        \"drug_canonical_smiles\": thapsigargin_smiles,\n",
    "        \"drug_moa_broad\": \"inhibitor/antagonist\",\n",
    "        \"drug_moa_fine\": \"ATPase inhibitor\",\n",
    "        \"drug_targets\": \"ATP2A1, ATP2A2, ATP2A3\",\n",
    "        \"drug_pubchem_cid\": \"5353993\"\n",
    "    },\n",
    "    \"3\": {\n",
    "        \"drug\": \"DMSO\",\n",
    "        \"drug_dose\": 0.0,\n",
    "        \"drug_canonical_smiles\": dmso_smiles,\n",
    "        \"drug_moa_broad\": None,\n",
    "        \"drug_moa_fine\": None,\n",
    "        \"drug_targets\": None,\n",
    "        \"drug_pubchem_cid\": \"679\"\n",
    "    }\n",
    "}\n",
    "compound_metadata = pd.DataFrame.from_dict(compound_metadata, orient=\"index\")\n",
    "\n",
    "# update compound metadata\n",
    "adamson_epistasis.obs[\n",
    "    [\"drug\", \"drug_dose\", \"drug_canonical_smiles\", \"drug_moa_broad\",\n",
    "     \"drug_moa_fine\", \"drug_targets\", \"drug_pubchem_cid\"]\n",
    "] = adamson_epistasis.obs_names.to_series().str.split(\"-\").apply(lambda x: compound_metadata.loc[x[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96bde695",
   "metadata": {},
   "source": [
    "**Add the rest of the obs metadata**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e1310f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "default_metadata = {\n",
    "    \"tissue\": \"cell line\",\n",
    "    \"cell_type\": \"lymphoblast\",\n",
    "    \"cell_line\": \"K562\",\n",
    "    \"disease\": \"chronic myelogenous leukemia\",\n",
    "    \"assay\": \"10x 3' v1\",\n",
    "    \"organism\": \"Homo sapiens\",\n",
    "    \"gene_perturbation_type\": \"CRISPRi\",\n",
    "    \"drug_dose_unit\": \"uM\",\n",
    "}\n",
    "extra_dtype_map = {\n",
    "    # Basic cell and read metadata\n",
    "    'cell_barcode': str,               # Unique identifier per cell\n",
    "    'guide identity': 'category',      # Guide RNA identifier (e.g., sgRNA ID)\n",
    "    'read count': int,                 # Raw read count for this cell\n",
    "    'UMI count': int,                  # Unique Molecular Identifier count\n",
    "    'coverage': float,                 # Sequencing depth or gene coverage\n",
    "    'good coverage': bool,             # QC filter (e.g., sufficient reads per cell)\n",
    "    'number of cells': int,            # Count aggregated per perturbation (optional per-cell)\n",
    "    'experiment': 'category',          # Experiment/batch label (e.g., plate ID)\n",
    "}\n",
    "\n",
    "# Apply to each AnnData object\n",
    "for ds in [adamson_perturb, adamson_epistasis, adamson_pilot]:\n",
    "    format_obs(ds, default_metadata=default_metadata, extra_dtype_map=extra_dtype_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afaddfe0",
   "metadata": {},
   "source": [
    "#### ii) var\n",
    "\n",
    "**Add gene_name**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d011284f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found local copy of fine_tune/datasets/genome/gencode.v32.primary_assembly.annotation.csv\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ../../ # cd to root of the project\n",
    "python fine_tune/scripts/genome.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ef6e6801",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = pd.read_csv(\n",
    "    data_path.joinpath(\"genome/gencode.v32.primary_assembly.annotation.csv\"),\n",
    "    index_col=\"ensembl_id\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8adf221f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/hhakem/projects/cellarium-ml/.venv/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/home/hhakem/projects/cellarium-ml/.venv/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/home/hhakem/projects/cellarium-ml/.venv/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "adamson_pilot = format_var(adamson_pilot, gene_name)\n",
    "adamson_epistasis = format_var(adamson_epistasis, gene_name)\n",
    "adamson_perturb = format_var(adamson_perturb, gene_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeb48a7d",
   "metadata": {},
   "source": [
    "### c) Save Adamson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "10a5e7f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adamson_pilot.write_h5ad(data_path.joinpath(\"anndata/adamson_pilot_GSM2406675_10X001.h5ad\"))\n",
    "adamson_epistasis.write_h5ad(data_path.joinpath(\"anndata/adamson_epistasis_GSM2406677_10X005.h5ad\"))\n",
    "adamson_perturb.write_h5ad(data_path.joinpath(\"anndata/adamson_perturb_GSM2406681_10X010.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64c30015",
   "metadata": {},
   "source": [
    "## 2) Norman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7b69dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths\n",
    "norman_dir = data_path.joinpath(\"raw/raw_norman\")\n",
    "\n",
    "# Load gene, barcode names, cell identities\n",
    "_genes = pd.read_csv(norman_dir / \"genes.tsv.gz\", header=None, sep=\"\\t\")\n",
    "barcodes = pd.read_csv(norman_dir / \"barcodes.tsv.gz\", header=None)\n",
    "cell_identities = pd.read_csv(norman_dir / \"cell_identities.csv.gz\", index_col=0)\n",
    "\n",
    "# Load matrix and filter out barcodes with no information:\n",
    "mask = barcodes.isin(cell_identities.index).to_numpy().ravel()\n",
    "matrix = mmread(norman_dir.joinpath(\"matrix.mtx.gz\")).tocsr().T[mask, :]\n",
    "\n",
    "# retain solely ensembl_id and not gene_name as we use custom gene_name\n",
    "genes = _genes.rename(columns={0: \"ensembl_id\"})[[\"ensembl_id\"]].set_index(\"ensembl_id\")\n",
    "\n",
    "# Build AnnData\n",
    "adata_norman = ad.AnnData(\n",
    "    X=matrix,\n",
    "    obs=cell_identities,\n",
    "    var=genes\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d41f730",
   "metadata": {},
   "source": [
    "### a) Retrieve correct perturbation\n",
    "\n",
    "The supplementary information provided in the paper need to be used to verify every `guide_identity` are within the expected 287 perturbations of the Norman dataset.\n",
    "\n",
    "- Table S7: transcriptional readouts for the 287 perturbations\n",
    "\n",
    "They cannot be retrieved by any of `curl`, `wget` (CLI based approach) or `pooch`, `urllib`, `request` (python based approach) due to a HTTP error 403.\n",
    "\n",
    "To bypass this, let's download the link from source on the local computer and then forward it to the server:\n",
    "\n",
    "- Click on the link to download the [tableS7](https://pmc.ncbi.nlm.nih.gov/articles/instance/6746554/bin/NIHMS1045467-supplement-Table_S7.xlsx)\n",
    "- Then forward it to your server using your preferred method:\n",
    "  Here using an `scp` command which is written similarly to an `ssh user@server-name`:\n",
    "  - Syntax: \n",
    "    ```bash\n",
    "    scp path-to-table-S7 user@server-name:output-path-to-save-table-S7\n",
    "    ```\n",
    "  - Actual Example:\n",
    "    ```bash\n",
    "    scp Downloads/NIHMS1045467-supplement-Table_S7.xlsx hhakem@hhakem-cellarium-t4:projects/cellarium-ml/fine_tune/datasets/raw/raw_norman\n",
    "    ```\n",
    "\n",
    "#### i) Process Table S7\n",
    "\n",
    "Let's verify the mapping between target and ensembl_id is correct according to the genome assembly annotation which is used to name ensembl_id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec43a079",
   "metadata": {},
   "outputs": [],
   "source": [
    "tableS7 = pd.read_excel(\n",
    "    norman_dir.joinpath(\"NIHMS1045467-supplement-Table_S7.xlsx\"),\n",
    "    header=21, # the top 21 rows are metadata\n",
    "    index_col=0\n",
    ").rename_axis(\"guide_identity\")\n",
    "\n",
    "\n",
    "gene_name = pd.read_csv(\n",
    "    data_path.joinpath(\"genome/gencode.v32.primary_assembly.annotation.csv\"),\n",
    "    index_col=\"ensembl_id\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c9bf91d",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_target_id = tableS7.reset_index()[[\"first_target\", \"first_id\", \"second_target\", \"second_id\"]]\n",
    "map_target_id = pd.concat(\n",
    "    [map_target_id[[\"first_target\", \"first_id\"]].rename(\n",
    "        columns={\"first_target\": \"Table S7\", \"first_id\": \"ensembl_id\"}),\n",
    "     map_target_id[[\"second_target\", \"second_id\"]].rename(\n",
    "         columns={\"second_target\": \"Table S7\", \"second_id\": \"ensembl_id\"})]\n",
    ").drop_duplicates().sort_values(by=\"Table S7\")\n",
    "\n",
    "# filter out NegCtrl\n",
    "map_target_id = map_target_id[~map_target_id[\"Table S7\"].str.contains(\"NegCtrl\")]\n",
    "\n",
    "if map_target_id.isna().any(axis=1).any():\n",
    "    raise ValueError(\n",
    "        \"There are non valid target - ensembl_id pair. See bellow:\\n\" \\\n",
    "        f\"{map_target_id[map_target_id.isna().any(axis=1)]}\"\n",
    "    )\n",
    "else:\n",
    "    map_target_id = map_target_id.set_index(\"ensembl_id\")\n",
    "    map_target_id[\"Assembly\"] = gene_name.loc[map_target_id.index][\"symbol\"]\n",
    "    print(\n",
    "        f\"There are {(mask := map_target_id['Table S7'] != map_target_id['Assembly']).sum()} \" \\\n",
    "        \"gene different between Table S7 and the genome assembly. Those are:\\n\" \\\n",
    "        f\"{map_target_id[mask]}\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f100e629",
   "metadata": {},
   "source": [
    "In those mismatch Table S7 use gene name that are not used anymore:\n",
    "\n",
    "| Ensembl_id      | Table S7    | Assembly | Gene Card Link                                                                                                                |\n",
    "|-----------------|-------------|----------|-------------------------------------------------------------------------------------------------------------------------------|\n",
    "| ENSG00000156030 | ELMSAN1     | MIDEAS   | https://www.genecards.org/cgi-bin/carddisp.pl?gene=MIDEAS                                                                     |\n",
    "| ENSG00000143674 | RP5-862P8.2 | MAP3K21  | https://www.genecards.org/cgi-bin/carddisp.pl?gene=MAP3K21 / https://thebiogrid.org/124089/summary/homo-sapiens/kiaa1804.html |\n",
    "\n",
    "Those are then renamed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97c044d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "tableS7 = tableS7.reset_index().map(\n",
    "    lambda x: x.replace(\"ELMSAN1\", \"MIDEAS\").replace(\"RP5-862P8.2\", \"MAP3K21\") if isinstance(x, str) else x\n",
    ").set_index(\"guide_identity\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3319ae9d",
   "metadata": {},
   "source": [
    "#### ii) Processing target_identity\n",
    "\n",
    "`guide_identity` is the form of: gene1_gene2__gene1_gene2. \n",
    "So it is double-double targeted gene (or ctrl) separated by a double underscore. Most of the time the pair of gene is the same but to make sure:\n",
    "- split at `__`\n",
    "- retrieve the unique pair\n",
    "\n",
    "Then those guide identity can be compared to **TableS7**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac672a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_guide = (adata_norman.obs[\"guide_identity\"]\n",
    " .drop_duplicates()\n",
    " # `dict.fromkeys()` ~ `set` but preserve order\n",
    " .str.split(\"__\").apply(lambda row: \"__\".join(list(dict.fromkeys(row))))\n",
    " .drop_duplicates()\n",
    " .to_frame()\n",
    " .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "unique_guide[[\"first_target\", \"second_target\"]] = unique_guide[\"guide_identity\"].str.split(\"_\", n=1, expand=True)\n",
    "unique_guide[[\"first_target\", \"second_target\"]] = unique_guide[[\"first_target\", \"second_target\"]].where(\n",
    "    unique_guide[\"guide_identity\"].str.count(\"_\") == 1,\n",
    "    unique_guide[[\"guide_identity\"]].values.repeat(2, axis=1),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# comparison to table S7\n",
    "unique_gene = pd.concat([unique_guide[\"first_target\"], unique_guide[\"second_target\"]]).drop_duplicates()\n",
    "unique_geneS7 = pd.concat([tableS7[\"first_target\"], tableS7[\"second_target\"]]).drop_duplicates()\n",
    "print(\"Non matching gene target in the norman `guide_identity` col compared to TableS7:\\n\"\\\n",
    "      f\"{unique_gene[~unique_gene.isin(unique_geneS7)]}\\n\\n\"\n",
    "      \"Non matching gene target in the TableS7 compared to `guide_identity` col:\\n\"\\\n",
    "      f\"{unique_geneS7[~unique_geneS7.isin(unique_gene)]}\\n\"\n",
    "      )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3167f09e",
   "metadata": {},
   "source": [
    "First: \n",
    "- There are remaining double underscore separated guide. They haven't been separated because at the \"__\" split, the left portion was different from the right one due to the addition of a end **tag** like `_1` or `_2`. That is most likely formatting error. \n",
    "  - Those end tag should be removed.\n",
    "- `no_reads_found` indicate cells whose identity couldn't be retrieved. Those should be filtered out. \n",
    "\n",
    "Second, there is a mismatch due to the usage of gene aliases: \n",
    "\n",
    "\n",
    "| guide_identity | table S7 | Gene Card Link                                             |\n",
    "|----------------|----------|------------------------------------------------------------|\n",
    "| C19orf26       | CBARP    | https://www.genecards.org/cgi-bin/carddisp.pl?gene=CBARP   |\n",
    "| C3orf72        | FOXL2NB  | https://www.genecards.org/cgi-bin/carddisp.pl?gene=FOXL2NB |\n",
    "| ELMSAN1        | MIDEAS   | https://www.genecards.org/cgi-bin/carddisp.pl?gene=MIDEAS  |\n",
    "| KIAA1804       | MAP3K21  | https://www.genecards.org/cgi-bin/carddisp.pl?gene=MAP3K21 |\n",
    "\n",
    "[RHOXF2](https://www.genecards.org/cgi-bin/carddisp.pl?gene=RHOXF2) and [RHOXF2B](https://www.genecards.org/cgi-bin/carddisp.pl?gene=RHOXF2B) are however two distinct genes. The paper mentioning **RHOXF2B**, it will be used as the actual target.\n",
    "\n",
    "Everything put together, this give:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49147b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_norman = adata_norman[\n",
    "    ~adata_norman.obs[\"guide_identity\"].str.contains(\"no_reads_found\")\n",
    "].copy()\n",
    "replacements = {\n",
    "    \"RHOXF2\": \"RHOXF2B\",\n",
    "    \"C19orf26\" : \"CBARP\",\n",
    "    \"C3orf72\" : \"FOXL2NB\",\n",
    "    \"ELMSAN1\": \"MIDEAS\",\n",
    "    \"KIAA1804\": \"MAP3K21\",\n",
    "}\n",
    "adata_norman.obs[\"guide_identity\"] = (\n",
    "    adata_norman.obs[\"guide_identity\"]\n",
    "    .str.replace(r\"_\\d+$\", \"\", regex=True) # remove ending tag\n",
    "    .map(lambda x: reduce(lambda s, kv: s.replace(*kv), replacements.items(), x)) # replace gene aliases\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e6bf162",
   "metadata": {},
   "source": [
    "Re-running the above cell with the updated the `adata_norman` leads to no more mismatch, confirming the guide_identity has been correctly processed.\n",
    "\n",
    "### b) Formatting Norman\n",
    "\n",
    "Norman is processed similarly as for [Adamson](#b-formatting-adamson).\n",
    "\n",
    "#### i) obs\n",
    "\n",
    "**Rename control and Parse guide identity**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7070d8bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_norman.obs[\"gene_perturbed\"] = (\n",
    "    adata_norman\n",
    "    .obs[\"guide_identity\"]\n",
    "    .str.split(\"__\", expand=True)[0]\n",
    "    .str.replace(r\"NegCtrl\\d+\", \"NT\", regex=True)\n",
    "    .str.replace(\"_\", \"+\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff4c09de",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_norman.obs[\"tissue\"] = \"cell line\"\n",
    "adata_norman.obs[\"cell_type\"] = \"lymphoblast\"\n",
    "adata_norman.obs[\"cell_line\"] = \"K562\"\n",
    "adata_norman.obs[\"disease\"] = \"chronic myelogenous leukemia\"\n",
    "adata_norman.obs[\"perturbation_type\"] = \"CRISPRa\"\n",
    "adata_norman.obs[\"compound\"] = None\n",
    "adata_norman.obs[\"compound_target\"] = None\n",
    "adata_norman.obs[\"compound_moa\"] = None\n",
    "adata_norman.obs[\"compound_dose_uM\"] = None\n",
    "adata_norman.obs[\"organism\"] = \"Homo sapiens\"\n",
    "adata_norman.obs[\"assay\"] = \"10x 3' v2\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
